src/app/contextBroker/contextBroker.cpp

```diff
@@ -772,12 +777,14 @@ static const char* validLogLevels[] =
 
 
 #define API_V2                                                                                         \
-  { "GET",    EPS,          EPS_COMPS_V2,         ENT_COMPS_WORD,          entryPointsTreat         }, \
-  { "*",      EPS,          EPS_COMPS_V2,         ENT_COMPS_WORD,          badVerbGetOnly           }, \
+  { "GET",     EPS,          EPS_COMPS_V2,        ENT_COMPS_WORD,          entryPointsTreat         }, \
+  { "OPTIONS", EPS,          EPS_COMPS_V2,        ENT_COMPS_WORD,          optionsGetOnly           }, \
```

Comment:
Service routines are named as VERB+'what'.
The 'what' in this case (/v2) is not very clear, but perhaps we can use 'entrypoint' as the previous service uses it.
[ Or v2root ... ? ]
The name of the service routine would then be `optionsEntrypoint`  (also, `entryPointsTreat` should be renamed to `getEntrypoint`, but not in this PR, of course).

Reply:
We had a [short discussion](https://github.com/telefonicaid/fiware-orion/issues/501#issuecomment-330555936) with @fgalan about this in #501. 

What we do for options requests is independent of the service logic, we are just pushing the necessary headers to the response. Therefore, to limit the number of files created, a similar approach as in badVerb... handlers was used.

I agree that using the form "optionsEntryPoint" is semantically better but this means we have to duplicate the same handler more than 10 times. For example, for every resource that allows GET and POST we need to create the same handler with a different name all of which are doing the same operation pushing "Access-Control-Allow-Methods: GET, POST" header to the response.

What do you think?

Reply:
So, if it is one single service routine, then we should find a generic name for it. 'options' is perhaps *too* generic, what about 'optionsTreat'. We've used Treat as function suffix in other places (for service routines)

Reply:
Let's see if I finally have understood this correctly (I have no prior knowledge of CORS, I'm afraid).

Each and every URL that the broker supports will be equipped with the OPTIONS method and only return the list of methods that the URL supports?
Which would mean that we need to add OPTIONS for the "Allow" header in all bad verb functions,
Funny, the info that you need with the OPTIONS verb is already there, thanks to the badVerb routines ...
Basically we have to copy all badVerb routines and make a minimal change (Allow => Access-Control-Allow-Methods).

[ We could even use the badVerb routines and depending on ciP->method do one thing or another ... ]

Reply:
> [ We could even use the badVerb routines and depending on ciP->method do one thing or another ... ]

I prefer to have a new method for that, given that bad verb processing and options verb processing are semantically different. In this line, the suggested name `optionsTreat` sounds good, although, if I'm understanding correctly, what we have at the end it not a single method covering all URLs, but several ones, covering each one a set of URLs. Thus, only a `optionsTreat` method probably is not enougth for all URL, but a small set of them (maybe using all them the `optionsTreat` prefix in the name) could be fine. This is a similar idea to current bad verb method, as far as I remember.

Reply:
Also related with this, I understand that OPTIONS processing make only sense in the case of CORS functionality (at least by the time being). Thus, the OPTIONS routes should be enabled only when CORS is activated. In other words, these `restServiceV[]` vector should include these entries only if `-enableCORS` (or whicheer name we use at the end for the CLI parameter) is being used.

We have done a similar "trick" with the `-ngsi9` switch (which disables all the NGSI9 routes in the V1 API). @kzangeli do you remember how we did that time? Otherwise, maybe we should look to old versions of the contextBroker.app (go to tag 0.24.0 in github).

Reply:
I should have added a more detailed description to this PR, I am sorry if I have caused any confusion. I'll try to make it a bit more clear for the community and for us :)

As discussed with Fermin in #501, the idea was to handle the issue in several PRs and this PR only implements the addition of the new switch, two `OPTIONS` handlers for `/v2` and `/v2/entities` routes and finally their tests.  In short, this PR is an implementation of CORS functionality for only two resources.

So in the short term, the goal is to cover all the `/v2` routes. There are 12 `/v2` routes and in total, there are 8 different sets of allowed methods.

* `GET`
* `POST`
* `GET POST`
* `GET PUT`
* `GET DELETE`
* `GET PUT DELETE`
* `GET DELETE PATCH`
* `GET POST PUT PATCH DELETE`

============================================================

If the ultimate goal is to support CORS **only for the /v2 routes**, I believe having 8 or 12 different handlers for OPTIONS requests is not a huge difference. So in this case, I think we can just create 12 routines dedicated to each route.

But if the ultimate goal is to support CORS for **both /v1 and /v2 routes**, I think it is better to follow the same logic in badVerb routines.

My question is, which one are we aiming for?

Reply:
I think the first approach (8-12 handlers depending on the allowed verbs in a given URL) is the best one. 

Moreover, I think the idea could be also applied to NGSIv1 in the future as for NGSIv1 is basically the same (in this case, most of the operations use only `POST`, so the handler for "only POST" will be used). However, let's put aside NGSIv1 by the moment. It is an almost-deprecated API which probably doesn't deserve to put any new functionality effort on it (only bugfixing).

Reply:
I think the initial idea of "copying" badVerb service routines is the way to go.
We don't need various functions dpimg the exact same thing.

Actually, there is another option here ...
Let's see what you think.
We could implement only one service routine, but a smarter one, that looks up the URL path in the ReatService vector and saves the verbs that it finds.I thought about changing the badVerb like this ... a few years ago ... Perhaps now it's time. It would be valid for both options and badVerb

Reply:
@McMutton please hold on at this point. I seems @kzangeli and me are proposing opposite alternatives :) So, we'll discuss and provide a conclusive alternative.

Reply:
After @kzangeli and @fgalan discusssion, we thenk the option of having 8-12 handlers depending on the allowed verbs in a given URL is the best one.

Handler name should follow the same criteria than bad verb, i.e. including in the name of the methods to be covered, e.g: `optionsGetOnly`, `optionsGetPostOnly`, etc.... as you are already doing as far as I see in this PR :)

Reply:
NTC?

(Not sure, but it seems the PR is already aligned with the conclusions to this comment threads, so no additional change needs to be done)

Reply:
Yes, NTC

Reply:
Yes, NTC